### ISSUE TITLE
Dynamic link not working with facebook

### ISSUE THREAD

**Issue by tabebqena** [firebase/firebase-android-sdk#916]
Facebook (the most popular social app.) uses special head meta tag to convert page to dynamic link.
Without this tag it opens the link in the 'inAppBrowser'.
Is there is way go integrate firebase dynamic links with facebook.

` <meta property="fb:app_id" content="your_fb_app_id" />`

`<meta property="al:android:package" content="com.your.package" />`
`<meta property="al:android:app_name" content="YourAwesomeApp" />`
`<meta property="al:android:url" content="somescheme://XYZ" />   <!-- identifiable by your app, your app must handle this data scheme inside your manifest's intent-filter tag --!>`
`<meta property="al:web:should_fallback" content="true" />`
`<meta property="al:web:url" content="http://www.yourwebsite.com/" />`

**Comment by LuizTokuhara** [comment]
Same problem here too

**Comment by AhmedSaber0** [comment]
any updates ?

**Comment by catleader** [comment]
now I can't event post dynamic links on Facebook or share via Facebook Messenger app, anyone?

**Comment by DanGDroid** [comment]
commented on this status
https://github.com/firebase/firebase-android-sdk/issues/915#issuecomment-546951346

**Comment by goneale** [comment]
We've been received customer reports that our dynamic links had stopped working on Facebook (opens in a blank white screen) and this is currently occurring in iOS and Android. I would have thought it was a FB app issue, or recent changes to dynamic link processing on Google (.link)? As we haven't updated our codebases in the stores recently. 
Should I create a bug on https://github.com/firebase/firebase-ios-sdk if we are experiencing this on iOS too?
Hoping the Firebase team has some leads on this...

**Comment by Sunillabs108** [comment]
On Android, through facebook App, if i open a shared dynamic link, It is not working, 
In OnSuccess listener Callback, (PendingDynamicLinkData pendingDynamicLinkData)  
pendingDynamicLinkData is coming as Null, So not able to retrieve deep link in the App.
Kindly suggest or fix this issue, 

WorkAround: I could  make it work by changing the Facebook privacy settings for enabling to open the link externally. then i could get the desired results. 
But this is not the correct way , As we can not force user to change the privacy settings of facebook.

**Comment by amgros** [comment]
any updates. our firebase links are NOT working anymore...why?? 
what privacy settings on Facebook can be changed to allow our links to start working in again??

**Comment by nightbear1009** [comment]
hi 
is there any update about this issue? our traffic comes from facebook 
this make us re-consider about using firebase dynamic link

**Comment by tabebqena** [comment]
as a temporary workaround, I suggest to use this [answer]https://stackoverflow.com/a/37834132)

**Comment by joshwolff1** [comment]
Same here

**Comment by joshwolff1** [comment]
I posted an answer that "sort of" fixes this here: https://stackoverflow.com/questions/41217810/firebase-dynamic-links-is-not-launching-my-app-in-specific-situation/59417981#59417981

Basically, remove the efr=1 parameter, which is what a developer suggested here: https://github.com/firebase/quickstart-ios/issues/798

**Comment by gpribi** [comment]
Same here! 
Found that when the activity handler is called from Facebook, the intent.getData() brings the right URL, but there's an "extra" missing called "com.google.firebase.dynamiclinks.DYNAMIC_LINK_DATA"

It's a byte[296] object with an unknown information, present when I use the link without Facebook.

**Comment by karapetyan** [comment]
Same here!

**Comment by KunalKathrotia** [comment]
Is there any update? dynamic links are not redirecting to the app when opening in facebook. works just fine everywhere else. what's the problem with facebook?

**Comment by mikehardy** [comment]
@gpribi @karapetyan @KunalKathrotia have you tried the er1 workaround from @joshwolff1 just above in https://github.com/firebase/firebase-android-sdk/issues/916#issuecomment-567716516

**Comment by KunalKathrotia** [comment]
@mikehardy what is efr and where can i find that parameter or from where do i've to remove that parameter as i am not able to find it.

**Comment by joshwolff1** [comment]
It confirmed does work for me.

**Comment by joshwolff1** [comment]
Here is a video of it working: https://youtu.be/KLAH39BTwmw

Here is the link I used: https://spontit.page.link/?link=https%3A//spontit.com/action/emilyslist/1577502216%3Fsc%3Dfb%26scuser%3Djosh-wolff&ibi=jw1.Spontit&isi=1448318683&efr=1

Here is the link with the debug parameter so that you can inspect it: https://spontit.page.link/?link=https%3A//spontit.com/action/emilyslist/1577502216%3Fsc%3Dfb%26scuser%3Djosh-wolff&ibi=jw1.Spontit&isi=1448318683&efr=1&d=1

iOS 13.3, Facebook Version 251.0

**Comment by joshwolff1** [comment]
One important note about the video I just took was that it did not open the preview page! I do not know why it did not show it. It has shown it in the past when I opened the link in the Facebook app. Either way, the App Store opened as expected.

**Comment by KunalKathrotia** [comment]
@joshwolff1 did you manually add efr=1 parameter in the original link before making it dynamic using firebase sdk?

**Comment by joshwolff1** [comment]
I manually constructed all of the links I have used. I have not attempted using the SDK.

**Comment by KunalKathrotia** [comment]
But we don't want to expose what the link really contains like that so we are using sdk to generate the dynamic link which will build short link `.buildShortDynamicLink(ShortDynamicLink.Suffix.SHORT)` and as far as the efr=1 is concerned we are already doing that as in above comment using that piece of code, in that `setForcedRedirectEnabled(true)` already sets efr=1 but it doesn't fix this issue

**Comment by joshwolff1** [comment]
Not familiar with the SDK, but can you manually construct it and then make it shorter? I know that you can do that from the console.

**Comment by KunalKathrotia** [comment]
we are generating the short links within our apps(android and ios apps) before sharing on any other platforms

**Comment by joshwolff1** [comment]
By "manual" I mean programmatically without the SDK (not literally manually).

If you can do that, then can you shorten it using the Firebase SDK? That's what I mean. I bet you can do that because you can do that from the console.

**Comment by KunalKathrotia** [comment]
i know what do you mean by manually, by following [this ](https://firebase.google.com/docs/dynamic-links/create-manually)i can generate a dynamic link, but what difference does it make because ultimately what i will be doing by generating that manually is same as generating the link using the sdk.

**Comment by joshwolff1** [comment]
It seems like there could be a bug. It could be in any part of the code. By manually constructing the link, you are removing the possibility that the bug is in the dynamic link creation part of the code. It's just a way to debug. I do not know what the cause is for your issue.

Try manually constructing, then using the console, and seeing if that works.

You have to try a lot of things and then something might work.

**Comment by KunalKathrotia** [comment]
Ok @joshwolff1 i'll also try that than.

**Comment by nightbear1009** [comment]
The efr works fine on ios but not work on android please help us

**Comment by joshwolff1** [comment]
> The efr works fine on ios but not work on android please help us

I have not tried this on Android. We have a different system set-up for that.

**Comment by KunalKathrotia** [comment]
The manually constructing dynamic link workaround is not working on android, but working on ios. So dynamically(using firebase sdks)/manually generated links shared on facebook when pressed it should redirect to app store/play store Or apps, which is not happening right now.

**Comment by KunalKathrotia** [comment]
The firebase dynamic links are not opening the app or app store based on the app's availability when clicked from skype too.

**Comment by rsevil** [comment]
Hi, I recently had to integrate dynamic links to one of my clients apps, and I got lots of trouble because of deeplinks not working on facebook+android_app.

As I couldn't move to branch.io because of princing reasons, I had to work it out on a different way so I read this issue on stackoverflow: https://stackoverflow.com/a/41592330. (more or less it says to put a proxy link between the client and the dynamic link itself where you should put the necessary facebook applinks metadata, and a redirect to the dynamic link itself for other types of clients)

I tried it with a hardcoded link and it worked very well (it opened the app instantly), so then I found out that I only needed to send via query string the link ID and another info to build some other metadata like og:title, og:description and og:image for the link preview to work well.

I ended with a very simple nodejs+express setup where I receive the dynamic link ID and return a HTML with all the meta-tags required by facebook applinks (android + iOS) and a redirect script (window.location = <dynamic_link>) for other clients to be used.

The best part is that now on facebook the app opens directly without the preview page in both platforms and the link works on all browsers via the script redirect.

Here is the gist: https://gist.github.com/rsevil/377690cef7bde5fe6825e3d248af035d

**Comment by mikehardy** [comment]
fascinating. Nicely done. Seems like this sort of thing could be a firebase function even? Anything you do in node/express essentially can be a firebase function, then everyone here (firebase users) can stay serverless

**Comment by rsevil** [comment]
Yes @mikehaney24 that could be done without problem, I didn't knew about firebase functions, but as I read the documentation I think it works that way too.

I didn't think about using serverless because I didn't had another custom domain to use and I used a site that was already being served under nodejs.

Anyways, this solution is more like a patch to what firebase dynamic links is lacking in functionality, and this should be working out of the box without a problem.

**Comment by mikehaney24** [comment]
@mikehardy was probably the intended mention

**Comment by FakhruddinAbdi** [comment]
@rsevil could you please elaborate on your solution.
I have the same problem, I tried your solution but no luck.
If you can give a sample link from your implementation or if you can check [this link](https://harzaan.com/post/-LzlsgVPEY-ZqCYMbrhT/) and let me know what i missed

**Comment by FakhruddinAbdi** [comment]
Another question is why you need dynamic link then? 
You are returning an HTML page, you can do it inside the original page which is linked in the dynamic link already.

**Comment by rsevil** [comment]
Hi @FakhruddinAbdi, I kept using dynamic links (redirecting to them) because it keeps me from handling all the logic needed of opening the store and all that sort of things.

One thing that I noted on the link you posted is that you were pointing to a custom scheme of your own, like if inside the "al:android:url" meta property you're trying to redirect to the deeplink. Instead there I'm using the url from a dynamic link that firebase is giving me, for example: dummydomain.page.link/aocjasocajc.

In the gist I have put everything I got and did to make it work.

Hope that it helps

**Comment by FakhruddinAbdi** [comment]
> al:android:url

Could you please give me your working link to check the details? @rsevil

**Comment by Aung-Myint-Thein** [comment]
I did some investigation and this is what I found out so far.

**TLDR: Firebase Dynamic Link SDK might be having an issue. Facebook In-App Mobile Browser for Android is definitely having issue in redirecting in my experience.**

**Scenario 1. Long link - Android working, iOS partially broken.**

I create a dynamic link manually. So, it is a long link such as "https://myanlearn.page.link/?link=https://www.myanlearn.com/course/607138951&apn=com.myanlearn_for_students&isi=1441239224&ibi=com.myanlearn.myanlearnforstudents". Posted the link on my Facebook Feed. When I open it in Android, my app is opened immediately and deep link is working as intended. When I open in iOS, the app is opened but deep link is not working. Surprisingly, no preview page is loaded in both iOS & Android.

**Scenario 2. Short link - Android (FB Mobile Browser) not working, iOS working.**

Then i create a dynamic link through REST API and got a link such as "https://myanlearn.page.link/vVp9eoQHnMSBLWxA6". Posted the link on my Facebook Feed. When I opened it in Android, Facebook's Mobile Browser tried to open my link and it opens "Play Store" page (if no fall back link `androidFallbackLink` is provided) OR opens fall back web link (if `androidFallbackLink` is provided). When I open in iOS, the preview page is loaded. Then my app is opened and deep link is working as intended. I tested the short link in all the other apps in Android and they are all working fine. So, I guess the way FB Mobile Browser is handling the link is broken somehow OR Firebase's detecting + directing is broken somehow in FB Mobile Browser.

Tested iOS: iPhone 6, iPhone 11
Android: OnePlus 2 (Android 6.0), Mi F1 Pocco (Android 9.0)

Hope this helps to find the issue and hopefully the fix...

**Comment by Aung-Myint-Thein** [comment]
After more investigation, I found out that the issue is in Firebase SDK to generate dynamic link short links. If there are `socialMetaTagInfo` data, the link is going to be broken and Facebook Mobile Browser will open web page. However, if there is no data for `socialMetaTagInfo` section, the link will work as intended. 

However, make sure your SHA256 fingerprints are in Firebase Console. You can get both your dev fingerprint and Google Play Store's key in **Play Store > Release Management > App Signing** and put them in **Firebase > Project Setting > Android Apps > SHA Fingerprints**.

Hope this helps!

**Comment by Aung-Myint-Thein** [comment]
Actually, I found out that 

without social meta tag, iOS won't work but Android will work. (eg: https://myanlearn.page.link/Cg29UhebgBRimiT48) This link will open my app in iOS without preview page (not sure why it skipped) but won't deep link into the content. Android is working fine.

with social meta tag, iOS will work but Android won't work. (eg: https://myanlearn.page.link/UF3RFFrALVGgPGYt6) Preview page is shown in iOS and deep linked into the content. Android won't open the app.

**Comment by sochychoeun** [comment]
Hi @samtstern my also has this issue, any status update?

**Comment by sanketsohaliya** [comment]
> Actually, I found out that
> 
> without social meta tag, iOS won't work but Android will work. (eg: https://myanlearn.page.link/Cg29UhebgBRimiT48) This link will open my app in iOS without preview page (not sure why it skipped) but won't deep link into the content. Android is working fine.
> 
> with social meta tag, iOS will work but Android won't work. (eg: https://myanlearn.page.link/UF3RFFrALVGgPGYt6) Preview page is shown in iOS and deep linked into the content. Android won't open the app.

For me short dynamic links on android does not work with or without social meta tags! It simply leads to the app on the play store but does not open the app even if it is installed!

I have also encountered another issue where dynamic links from other app on android only work if you have chrome browser open otherwise It simply leads to the app on the play store but does not open the app even if it is installed!

**Comment by armanhadifatah** [comment]
I have the same problem with dynamic link on facebook but it works fine on messenger,
and on Viber it does not show my image it shows firebase logo.

can anyone help ? 
   because i really need that my dynamic links to be opened from facebook but what i get it redirect user to firebase url and then ask user to open play store after that when clicking on play store icon it open the app.

**Comment by heshaShawky** [comment]
I'm using flutter and this issue still, Facebook opens the play store instead of the app and it's installed and working on WhatsApp fine, but most of the app marketing depends on Facebook so this is a huge deal. 

I'm talking about the Android version.

**Comment by aanalmehta** [comment]
I am also facing the same issue for Facebook and Skype.
Platform: Android

**Comment by superturboryan** [comment]
I was able to get it to work for iOS by adding the appropriate metadata to our website's main page's header (used as the dynamic link URL set in Dynamic Links editor) as referenced here https://developers.facebook.com/docs/applinks/metadata-reference/

**Comment by heshaShawky** [comment]
> I was able to get it to work for iOS by adding the appropriate metadata to our website's main page's header (used as the dynamic link URL set in Dynamic Links editor) as referenced here https://developers.facebook.com/docs/applinks/metadata-reference/

What about android? it redirects me to google play instead of the URL ( website ) so using metadata not helpful

**Comment by heshaShawky** [comment]
> I did some investigation and this is what I found out so far.
> 
> **TLDR: Firebase Dynamic Link SDK might be having an issue. Facebook In-App Mobile Browser for Android is definitely having issue in redirecting in my experience.**
> 
> **Scenario 1. Long link - Android working, iOS partially broken.**
> 
> I create a dynamic link manually. So, it is a long link such as "https://myanlearn.page.link/?link=https://www.myanlearn.com/course/607138951&apn=com.myanlearn_for_students&isi=1441239224&ibi=com.myanlearn.myanlearnforstudents". Posted the link on my Facebook Feed. When I open it in Android, my app is opened immediately and deep link is working as intended. When I open in iOS, the app is opened but deep link is not working. Surprisingly, no preview page is loaded in both iOS & Android.
> 
> **Scenario 2. Short link - Android (FB Mobile Browser) not working, iOS working.**
> 
> Then i create a dynamic link through REST API and got a link such as "https://myanlearn.page.link/vVp9eoQHnMSBLWxA6". Posted the link on my Facebook Feed. When I opened it in Android, Facebook's Mobile Browser tried to open my link and it opens "Play Store" page (if no fall back link `androidFallbackLink` is provided) OR opens fall back web link (if `androidFallbackLink` is provided). When I open in iOS, the preview page is loaded. Then my app is opened and deep link is working as intended. I tested the short link in all the other apps in Android and they are all working fine. So, I guess the way FB Mobile Browser is handling the link is broken somehow OR Firebase's detecting + directing is broken somehow in FB Mobile Browser.
> 
> Tested iOS: iPhone 6, iPhone 11
> Android: OnePlus 2 (Android 6.0), Mi F1 Pocco (Android 9.0)
> 
> Hope this helps to find the issue and hopefully the fix...

I tested all of those scenarios and nothing works with me ( using flutter ) on an android device.

**Comment by Abhinav2807** [comment]
> I did some investigation and this is what I found out so far.
> 
> **TLDR: Firebase Dynamic Link SDK might be having an issue. Facebook In-App Mobile Browser for Android is definitely having issue in redirecting in my experience.**
> 
> **Scenario 1. Long link - Android working, iOS partially broken.**
> 
> I create a dynamic link manually. So, it is a long link such as "https://myanlearn.page.link/?link=https://www.myanlearn.com/course/607138951&apn=com.myanlearn_for_students&isi=1441239224&ibi=com.myanlearn.myanlearnforstudents". Posted the link on my Facebook Feed. When I open it in Android, my app is opened immediately and deep link is working as intended. When I open in iOS, the app is opened but deep link is not working. Surprisingly, no preview page is loaded in both iOS & Android.
> 
> **Scenario 2. Short link - Android (FB Mobile Browser) not working, iOS working.**
> 
> Then i create a dynamic link through REST API and got a link such as "https://myanlearn.page.link/vVp9eoQHnMSBLWxA6". Posted the link on my Facebook Feed. When I opened it in Android, Facebook's Mobile Browser tried to open my link and it opens "Play Store" page (if no fall back link `androidFallbackLink` is provided) OR opens fall back web link (if `androidFallbackLink` is provided). When I open in iOS, the preview page is loaded. Then my app is opened and deep link is working as intended. I tested the short link in all the other apps in Android and they are all working fine. So, I guess the way FB Mobile Browser is handling the link is broken somehow OR Firebase's detecting + directing is broken somehow in FB Mobile Browser.
> 
> Tested iOS: iPhone 6, iPhone 11
> Android: OnePlus 2 (Android 6.0), Mi F1 Pocco (Android 9.0)
> 
> Hope this helps to find the issue and hopefully the fix...

I tried everything mentioned above , but nothing seems to work for me . If someone has a solution please help .

**Comment by hiendc170196** [comment]
> I did some investigation and this is what I found out so far.
> 
> **TLDR: Firebase Dynamic Link SDK might be having an issue. Facebook In-App Mobile Browser for Android is definitely having issue in redirecting in my experience.**
> 
> **Scenario 1. Long link - Android working, iOS partially broken.**
> 
> I create a dynamic link manually. So, it is a long link such as "https://myanlearn.page.link/?link=https://www.myanlearn.com/course/607138951&apn=com.myanlearn_for_students&isi=1441239224&ibi=com.myanlearn.myanlearnforstudents". Posted the link on my Facebook Feed. When I open it in Android, my app is opened immediately and deep link is working as intended. When I open in iOS, the app is opened but deep link is not working. Surprisingly, no preview page is loaded in both iOS & Android.
> 
> **Scenario 2. Short link - Android (FB Mobile Browser) not working, iOS working.**
> 
> Then i create a dynamic link through REST API and got a link such as "https://myanlearn.page.link/vVp9eoQHnMSBLWxA6". Posted the link on my Facebook Feed. When I opened it in Android, Facebook's Mobile Browser tried to open my link and it opens "Play Store" page (if no fall back link `androidFallbackLink` is provided) OR opens fall back web link (if `androidFallbackLink` is provided). When I open in iOS, the preview page is loaded. Then my app is opened and deep link is working as intended. I tested the short link in all the other apps in Android and they are all working fine. So, I guess the way FB Mobile Browser is handling the link is broken somehow OR Firebase's detecting + directing is broken somehow in FB Mobile Browser.
> 
> Tested iOS: iPhone 6, iPhone 11
> Android: OnePlus 2 (Android 6.0), Mi F1 Pocco (Android 9.0)
> 
> Hope this helps to find the issue and hopefully the fix...

How you configure the link in scenario 1 to open in facebook's feed without set "Open external link" in Setting? I tried configure my own dynamic link manually on Firebase Console but when I click it in Facebook's Feed ,it redirected to Play Store instead of my app (I tested on Android and don't open external link on Facebook). May you explain your own way? I would appreciate it

**Comment by hiendc170196** [comment]
Now I facing two issues:
1.When I paste the link in progress bar on Chrome, it redirect to Play Store even thought I have already installed app
2.When I click my own link on Facebook's Feed,it didn't redirect to correct screen that I want to move
From my standpoint,it probably the Firebase's issue because I try another social networks such as Instagram, Skype, Messenger,Zalo it all work

**Comment by Significantinfotech2020** [comment]
> Now I facing two issues:
> 1.When I paste the link in progress bar on Chrome, it redirect to Play Store even thought I have already installed app
> 2.When I click my own link on Facebook's Feed,it didn't redirect to correct screen that I want to move
> From my standpoint,it probably the Firebase's issue because I try another social networks such as Instagram, Skype, Messenger,Zalo it all work

I have same issue. If any one got solution, then please suggest us.

**Comment by Spinthecode** [comment]
Hi any update? thank you!

**Comment by MaxJadav** [comment]
Any update on this? Fetch same issue in FB, skype, Chrome. If any one got solution, then please suggest us.

**Comment by JurajBegovac** [comment]
Any update? E.g. - same link is working when launched from Facebook inside chrome app and not working when launched from Facebook mobile app. When launched from Facebook app it opens play store no matter that I already have the app installed.

**Comment by bchanthavong-ss** [comment]
Running into the same issue on Android.  Posting a short link in the Facebook app always goes to the Play Store even if the app is installed.  If the app is not installed, when installing it, it does not have the deep link.

**Comment by xuancelestial** [comment]
Same issue here

**Comment by enginebai** [comment]
Any update for this, it's still broken for dynamic link `20.1.0`.

**Comment by AndrewMorsillo** [comment]
@samtstern this still seems to be an issue over 1.5 years after it was reported, is there anybody that should be tagged to look into this?

**Comment by githubbla** [comment]
one more month? any way to fix this?

**Comment by BurningAXE** [comment]
@samtstern Can we at least get some response to know if anybody is doing anything to fix this issue? 
It's very disappointing to see a 2-years old issue with a lot of affected users and no progress whatsoever!

**Comment by eagarcias** [comment]
@samtstern any news?

**Comment by BurningAXE** [comment]
Guys, since nobody from the Firebase team reacts to this two other issues about Dynamic Links not working (properly) in Facebook app, I contacted Firebase through their support. This is the response I got:

> Just wanted to let you know that we are aware that Dynamic Links don't work well with Facebook. Unfortunately, we don't have a workaround for this issue, as it's up to Facebook to determine how URLs are handled inside their app. The only suggestion I can provide would be to ask your users to open your links from a non-Facebook browser. Hope you understand.

So, don't expect any fixes from Firebase and deal with the issue accordingly...

**Comment by eagarcias** [comment]
@BurningAXE thanks

**Comment by sakher** [comment]
Same in linkedin, as well as in Chrome desktop when you emulate mobile, it renders dynamic links completely useless to be honest! any workarounds?

**Comment by BurningAXE** [comment]
Well, I switched to Branch.io for dynamic links and it works like charm everywhere!

**Comment by aanalmehta** [comment]
Branch IO is the best option so far I have used. I have observed that, there are few issues with dynamic links with Skype and Facebook.
So, the best option is Branch IO.

**Comment by dngconsulting** [comment]
We are in 2022 and same issue. What I don't understand is that I can easily deeplink google photos or any other google app in Facebook without any issue. Looks like google doesn't use dynamic links for their own apps...

**Comment by mikehardy** [comment]
@dngconsulting that may be true for some things but I think the Google Maps "copy to clipboard" version of share link is a deep link based on the behavior I see when I use it. Have you tried sharing those on facebook ? Might make an interesting test of your assertion

**Comment by dngconsulting** [comment]
Interesting @mikehardy . When one try with Google Photo it opens up the facebook web browser with a preview from the Google Photo website showing a button "Open in the native App" (btw curious to see how a web page can detect whether a native app is installed or not). For Google Map, same issue, it opens up the Google Map Web page preview (no deeplink). Same with Google Drive, the webview is always proposed with no way to deep link inside the FB browser. 
This approach is interesting, I could propose an intermediary page stating that the user has to open a real browser but by default on android, dynamic links will systematically propose the AppStore default page. Is there any way to override this behaviour (please don't propose me to host my own endpoint and play with user agent ;-)....).

**Comment by MRTHAKER** [comment]
i am having same issue but interesting observation i made & found that only deffered link is broken for me from the testing tool facebook provides : https://developers.facebook.com/tools/app-ads-helper/
here normal link works on android but if i send deffered it always open play store & same thing on facebook ads.
if anyone has any hotfix please share.

**Comment by kmjadeja** [comment]
Guys you just need to skip the social meta details (Step#5) when you are creating a dynamic link from the Firebase console.

## You just need to remove this meta if you are creating a dynamic link for Facebook AD.

If you are creating a dynamic link by using REST API then there you just need to remove ```socialMetaTagInfo``` or you can skip the complete ```dynamicLinkInfo```.

I've attached the image for you reference you guys can check.
----

**Generate Dynamic Link Using Firebase Console**

![Firebase Dynamic Link - Console](https://user-images.githubusercontent.com/31174014/155329077-59fca194-e4d5-4398-8c7a-596829250000.png)

**Generate Dynamic Link Using REST API**

![Firebase Dynamic Link - REST API](https://user-images.githubusercontent.com/31174014/155328721-2d56c450-934e-46ef-821d-1660eac5abe4.png)


⚠️ **Note** : 
- This is only for the Facebook AD.
- You need to create a new link if you have already created, due to caching issue.


Give 👍  or 🎉 if this works for you. 😉
Thanks & Happy Coding 🍻

**Comment by dngconsulting** [comment]
All the links generated in my app are dynamic in the code with the firebase SDK. I need to pass a parameter to every links (id=XX) to navigate with contextual deeplink. How to conditionally put the social meta tags regarding whether the link is clicked on Facebook or in other apps ?  How would you do that with your way @kmjadeja ?

**Comment by kmjadeja** [comment]
> All the links generated in my app are dynamic in the code with the firebase SDK. I need to pass a parameter to every links (id=XX) to navigate with contextual deeplink. How to conditionally put the social meta tags regarding whether the link is clicked on Facebook or in other apps ? How would you do that with your way @kmjadeja ?

Hey @dngconsulting,

We create the dynamic links through REST API and we give a separate section where back-end can create different links for social media AD and other marketing.

For our app we have a team who run campaign over social-media, so for that specific user type backed will create a link without meta tags.

**Comment by dngconsulting** [comment]
Ok, this behaviour suit your specific needs but not the common case which consist to use the same link for all platforms. Our users choose to share on facebook/whatsapp or any other platform, we simply can't propose specific links for facebook and other for whatsapp/instagram. Those links should work on facebook with social meta tags or not. But I understand that it's more a facebook issue than a firebase one.

**Comment by kmjadeja** [comment]
Hello @dngconsulting,

I've worked on this issue and found that many people are facing the same issue. So I found one solution which will work with your facebook or social media and campaigns.  I've used the concept of firebase & universal link. I've used this technique in my live projects and tested for a 1 Month and we didn't found any issue. Given solutions is tested and trusted by me and my team on live projects.

I've written a blog with code snippets so you guys are check the details and share your feedback.
If you have any problems in implementation do let me know, I'll try to help out.

Link : [Google Dynamic Links](https://www.codezma.com/php/google-dynamic-links/#dynamic-links-with-facebook-ad)


#help #firebase #facebook #codezma #blog #code #snippets
